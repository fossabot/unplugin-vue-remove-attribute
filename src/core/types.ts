import type { FilterPattern } from '@rollup/pluginutils'

export interface Options {
  include?: FilterPattern
  exclude?: FilterPattern

  enabled?: boolean

  remove?: FilterPattern
}
