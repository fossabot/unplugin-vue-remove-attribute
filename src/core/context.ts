import { Options } from './types'
import { createFilter } from '@rollup/pluginutils'

export class Context {
  enable = true
  buildOnly = true

  options: Options

  attrsFilter: (id: unknown)=> boolean

  constructor(options: Options = {}) {

    if (options.enabled === false) this.enable = false
    if (options.buildOnly === false) this.buildOnly = false

    this.options = options
    this.attrsFilter = createFilter(options.remove || [/^data-test/])

  }
}
