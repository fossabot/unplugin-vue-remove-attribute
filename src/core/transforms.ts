import { parse } from '@vue/compiler-sfc'
import { Context } from './context'
import { getElements, NodeTypes } from './utils/ast'

export const vue = (code: string, ctx: Context): string => {
  const result = parse(code)

  if (!result.descriptor.template) return code

  let deleteLocs = []

  for (const child of getElements(result.descriptor.template.ast)) {
    if (child.type !== 1) continue

    for (const prop of child.props) {

      if (prop.type === NodeTypes.ATTRIBUTE) {
        if (!ctx.attrsFilter(prop.name)) continue

        deleteLocs.push({
          start: prop.loc.start.offset,
          end: prop.loc.end.offset,
        })
      } else {
        if (!prop.arg) continue
        let name
        if (prop.arg.type === NodeTypes.SIMPLE_EXPRESSION) {
          name = prop.arg.content
        } else {
          // TODO ?
        }
        if (!ctx.attrsFilter(name)) continue

        deleteLocs.push({
          start: prop.loc.start.offset,
          end: prop.loc.end.offset,
        })
      }
    }
  }

  deleteLocs = deleteLocs.sort((a, b) => b.start - a.start)

  for (const loc of deleteLocs) {
    code = code.substring(0, loc.start) + code.substring(loc.end)
  }

  return code
}
