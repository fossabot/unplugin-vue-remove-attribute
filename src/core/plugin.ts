import { createUnplugin } from 'unplugin'
import { Options } from './types'
import { createFilter } from '@rollup/pluginutils'
import { vue } from './transforms'
import { Context } from './context'

export const plugin = createUnplugin<Options>((options = {}) => {
  const filter = createFilter(
    options.include || [/\.vue$/, /\.vue\?vue/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/, /[\\/]\.nuxt[\\/]/]
  )

  const ctx = new Context(options)

  return {
    name: 'unplugin-vue-remove-attribute',
    enforce: 'pre',
    transformInclude(id): boolean {
      return filter(id)
    },
    transform(code): string {
      if (!ctx.enable) return code

      return vue(code, ctx)
    },
    vite: {
      config(config, { command }): void {
        if (command !== 'build') ctx.enable = false
      },
    },
  }
})

export default plugin
