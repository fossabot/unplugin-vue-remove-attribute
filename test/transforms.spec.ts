import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { vue } from '../src/core/transforms'
import { Context } from '../src/core/context'

describe('transform', () => {

  const root = resolve(__dirname)
  const fixtures = import.meta.glob('./fixtures/transform/*.vue')

  for (const file of Object.keys(fixtures)) {

    it(file, () => {
      const fixture = readFileSync(resolve(root, file), 'utf-8')
      const result = vue(fixture, new Context())

      expect(result).toMatchSnapshot()

      const result2 = vue(result, new Context())

      expect(result2).toMatch(result)
    })
  }

})
