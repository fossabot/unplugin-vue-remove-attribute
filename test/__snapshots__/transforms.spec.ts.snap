// Vitest Snapshot v1

exports[`transform > ./fixtures/transform/Attribute.vue 1`] = `
"<template>
  <div class=\\"hello\\"   />
</template>
"
`;

exports[`transform > ./fixtures/transform/Bind.vue 1`] = `
"<template>
  <div  />
</template>
"
`;

exports[`transform > ./fixtures/transform/Model.vue 1`] = `
"<script setup>
import { ref } from 'vue'

const model = ref('')
</script>

<template>
  <CustomComponent  />
</template>
"
`;

exports[`transform > ./fixtures/transform/NoChange.vue 1`] = `
"<template>
  <div class=\\"hello\\" />
  <div id=\\"123\\">
    <CustomComponent custom-attribute=\\"custom-attribute\\" />
    <button :type=\\"'submit'\\" :=\\"{}\\">
      hello world
    </button>
  </div>
</template>
"
`;

exports[`transform > ./fixtures/transform/ParseError.vue 1`] = `
"test
"
`;
